### ISSUE TITLE
LA OAuth2

### ISSUE THREAD

**Issue by SuruPa00** [BIDMCDigitalPsychiatry/LAMP-platform#440]
> is it possible to federate mindLAMP account management with Azure Active Directory. The reason we want to explore this option is because it could potentially save a lot of work and time for your development team in updating the password/credential requirements that have been identified. Since DMH has set rule around the Azure AD that meets auditor’s requirements, if the application can be federated with Azure AD, the following security changes can potentially be avoided. 

> 1.	Multifactor Authentication (MFA)
> 2.	Password Rule (8 characters long, at least one upper case, one lower case, one special characters)
> 3.	Password expires (password expires every 90 days, user must be prompt to select new password)
> 4.	Ability for user to reset their own password
> 5.	logon ID is suspended automatically after a predetermined number of invalid logon attempts.
> 6.	Users are required to change their password at initial log-in to meet the password rule. 
> 7.	The system times-out to prevent unauthorized access to unattended workstations. (After 15 minutes of inactivates)

**Comment by avaidyam** [comment]
A draft flowchart for the client-side operation is attached (more details coming later).
![image](https://user-images.githubusercontent.com/1361003/148802207-33b1971e-7c5f-4e20-af4f-68f9f25d557b.png)

**Comment by avaidyam** [comment]
@mcurbelo-orangeloops @falmeida-orangeloops Let's track discussion around this feature here in this issue! 😄

**Comment by mcurbelo-orangeloops** [comment]
@avaidyam 
Hey guys. Support for OAuth 2 is something that we are currently actively working on. Our goal is to support a wide variety of identity providers, and have our own one. The first milestone for this project is using Azure AD as a method of authentication. As of writing this, the Azure AD implementation is ongoing and mostly completed, targeting public release on June-July

One of the many things we've spoken about is expanding our current objectives from "Use Azure AD" to "Use Azure AD B2C" so that users have a wider variety of options to chose from when logging in.

As of this moment we're in the process of defining exactly how the B2C implementation of this is going to work, but we will try to keep this issue updated as more information becomes available.


Regarding B2C specifically:

The main doubt we have right now is that B2C seems to challenge one of the main assumptions that we have made for the Azure AD implementation.

The assumption is:

`Users that try to log in through an IdP will have a pre-created credential in the MindLAMP Systems. This credential will be email matched (Meaning we will match the IdP email with the MindLAMP email, and if no match, we just show an error).`

For B2C, does the same assumption stand ? Do you tell participants “Hey, what email will you try to signup with?” beforehand, then create a credential ?

As you know, the main reason we went with that assumption is because of the origin data. If a user first creates a credential through B2C we will not have an origin for it, which means it will need to be updated after the fact. So we need to create the necessary mechanisms for that.

**Comment by avaidyam** [comment]
Update on this feature: code review on all 3 PRs is completed and I believe we are ready to share for QA testing. Since the API changes won't affect the end product, I think we should do these two things in parallel.

**Comment by mcurbelo-orangeloops** [comment]
@avaidyam Regarding the issue we have received from LA County, this is what I have found so far.
So we need to do 2/3 things before they are ready to go

1. Unify the dashboard git branches. We currently have a feature/oauth and a feature/oauth-merge branch as a result of issues I ran into during a merge. They are using the first branch, which is a bit outdated. We should unify all the currently progress into the branch they are using so their next deployment works as expected.

2. We need to figure out a way for the dashboard to use the correct version of the LAMP-js library. This library does some of the heavy lifting for oauth to work (Mostly connection to the backend). The problem is that we don’t have a published version of this library with the oauth changes. They are using version "lamp-core": "^2022.3.8" which doesn’t include the necessary changes to correctly interface with the oauth-enabled backend.

3. They mentioned during the meeting that they were unsure as to how to create an admin account, and I think that was something we overlooked. The way the MindLAMP-AzureAD connection work is through email linking, so we will need to create a local admin credential during server boot for them.

**Comment by avaidyam** [comment]
@mcurbelo-orangeloops It should be possible to directly reference a Git repo for npm packages. (As we discussed on Slack.) ![image.png](https://user-images.githubusercontent.com/1361003/180470211-45bd4208-c647-4c14-b993-96a04f6e961f.png)

**Comment by mcurbelo-orangeloops** [comment]
Another issue I just ran into when setting up my local environment: The dashboard has this dependency "material-ui": "^0.20.2", which is a library that seems to have been deprecated ~4 years ago. When I tried to install the dashboard dependencies I saw this error. I was able to “fix it” by using npm install --force but we should probably migrate away from material-ui to material-ui/core and remove the former dependency entirely. I am not sure how much work that is going to be though as I don’t know what we’re using the dependency for

<img width="885" alt="Screen Shot 2022-07-22 at 11 58 38 AM" src="https://user-images.githubusercontent.com/42544460/180470357-813d5342-370e-43da-a916-5d5d14bfd349.png">

**Comment by mcurbelo-orangeloops** [comment]
> @mcurbelo-orangeloops It should be possible to directly reference a Git repo for npm packages. (As we discussed on Slack.) ![image.png](https://user-images.githubusercontent.com/1361003/180470211-45bd4208-c647-4c14-b993-96a04f6e961f.png)

Hey, we know it’s possible but wouldn’t recommend it. The branch might change or be deleted and it might result in unexpected issues down the line. It could be used as a stop-gap solution. We’re looking into github packages as a more long term one, trying to gauge the effort

**Comment by mcurbelo-orangeloops** [comment]
@avaidyam Hey, I think a good compromise is to depend from a tag instead of adding a dependency directly to a branch, which is by nature unstable. What do you think about doing that instead ?

**Comment by mcurbelo-orangeloops** [comment]
> They mentioned during the meeting that they were unsure as to how to create an admin account, and I think that was something we overlooked. The way the MindLAMP-AzureAD connection work is through email linking, so we will need to create a local admin credential during server boot for them.

I will create a new environment variable with an email that will create a credential on server boot, so that they can log in with an administrator account, and they can then configure that email to be whatever they want.

**Comment by avaidyam** [comment]
> depend from a tag instead of adding a dependency directly to a branch, which is by nature unstable

@mcurbelo-orangeloops I think because this is a temporary situation specific to the `feature/oauth` & co. branches across repositories, this is a good idea!

**Comment by mcurbelo-orangeloops** [comment]
We've added a new property to the environment.

`DB_ADMIN_USERNAME="admin@admin.com"` -> This property, when set & OAuth is enabled, will create admin credential without a secret key for use with third party log in.

**Comment by ertjlane** [comment]
I wanted to document a broad update of the project here. Currently there are two primary components of implementation. The first is [Personal Access Tokens for 2fa Oauth Support](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/738). This is currently in qa testing, with any feedback (if necessary) expected by the end of 03/24/23. 
Then there is [Loopback Oauth server for credentials support](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/739). An evaluation of the effort required for this is expected by April 1st, and from there a timeline can be outlined.
